lib/client/conntest/database.go

```diff
@@ -0,0 +1,394 @@
+// The following checkpoints are reported:
+// - database server for the requested database exists / the user's roles can access it
+// - the user can use the requested database user and database name (per their roles)
+// - the database is acessible and accepting connections from the database server
+// - the database has the database user and database name that was requested
+func (s *DatabaseConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
+	if req.ResourceKind != types.KindDatabase {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindDatabase, req.ResourceKind)
+	}
+
+	if req.DatabaseUser == "" {
+		return nil, trace.BadParameter("missing required parameter Database User")
+	}
+
+	if req.DatabaseName == "" {
+		return nil, trace.BadParameter("missing required parameter Database Name")
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(
+		connectionDiagnosticID,
+		map[string]string{},
+		types.ConnectionDiagnosticSpecV1{
+			// We start with a failed state so that we don't need to set it to each return statement once an error is returned.
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Lookup the Database Server that's proxying the requested Database.
+	listResourcesResponse, err := s.cfg.UserClient.ListResources(ctx, proto.ListResourcesRequest{
+		PredicateExpression: fmt.Sprintf(`name == "%s"`, req.ResourceName),
+		ResourceType:        types.KindDatabaseServer,
+		Limit:               1,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	databaseServers, err := types.ResourcesWithLabels(listResourcesResponse.Resources).AsDatabaseServers()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(databaseServers) == 0 {
+		connDiag, err := s.appendDiagnosticTrace(ctx,
+			connectionDiagnosticID,
+			types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+			"You are not authorized to access this Database. "+
+				"Ensure your role grants access by adding it to the 'db_labels' property. "+
+				"This can also happen when you don't have a Database Agent proxying the database - "+
+				"you can fix that by adding the database labels to the 'db_service.resources.labels' in 'teleport.yaml' file of the database agent.",
+			trace.NotFound("%s not found", req.ResourceName),
+		)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	databaseServer := databaseServers[0]
```

Comment:
I think it is worth including some database server info (e.g. host id/name) in the trace if we can. In HA setup, it is possible some agents can connect but others have problems.